source/Calamari.CloudAccounts/AzureOidcAccountExtensions.cs

```diff
@@ -16,36 +16,62 @@ public static async Task<ServiceClientCredentials> Credentials(this AzureOidcAcc
+        static async Task<string> GetAuthorizationToken(string tenantId, string applicationId, string token, string managementEndPoint, string activeDirectoryEndPoint, string aureEnvironment, string instanceDiscoveryUri, CancellationToken cancellationToken)
         {
             var authContext = GetOidcContextUri(string.IsNullOrEmpty(activeDirectoryEndPoint) ? "https://login.microsoftonline.com/" : activeDirectoryEndPoint, tenantId);
             Log.Verbose($"Authentication Context: {authContext}");
+            
+            var builder = ConfidentialClientApplicationBuilder.Create(applicationId).WithClientAssertion(token);
 
-            var app = ConfidentialClientApplicationBuilder.Create(applicationId)
-                                                          .WithClientAssertion(token)
-                                                          .WithAuthority(authContext)
-                                                          .Build();
+            if (!string.IsNullOrEmpty(instanceDiscoveryUri))
+            {
+                builder = builder.WithInstanceDiscoveryMetadata(new Uri(instanceDiscoveryUri))
+                                 .WithAuthority(authContext, false);
+            }
+            else
+            {
+                builder = builder.WithAuthority(authContext);
+            }
+
+            var app = builder.Build();
+
+            // Default values set on a per cloud basis on AzureOidcAccount, if managementEndPoint is set on the account /.default is required.
+            var scope = managementEndPoint == DefaultVariables.ResourceManagementEndpoint || string.IsNullOrEmpty(managementEndPoint)
+                ? AzureOidcAccount.GetDefaultScope(aureEnvironment)
+                : managementEndPoint.EndsWith(".default")
+                    ? managementEndPoint
+                    : $"{managementEndPoint}/.default";
 
-            var result = await app.AcquireTokenForClient(
-                                                         // Default values set on a per cloud basis on AzureOidcAccount, if managementEndPoint is set on the account /.default is required.
-                                                         new[] { managementEndPoint == DefaultVariables.ResourceManagementEndpoint || string.IsNullOrEmpty(managementEndPoint) ? AzureOidcAccount.GetDefaultScope(aureEnvironment) : managementEndPoint.EndsWith(".default") ? managementEndPoint : $"{managementEndPoint}/.default" })
+            var result = await app.AcquireTokenForClient(new[] { scope })
                                   .WithTenantId(tenantId)
                                   .ExecuteAsync(cancellationToken)
                                   .ConfigureAwait(false);
-            return result.AccessToken;
+
+            var accessToken = result.AccessToken;
+            if (string.IsNullOrEmpty(accessToken))
+            {
+                throw new Exception("Failed to get access token");
+            }
```

Comment:
Throwing to short circuit the auth flow. Otherwise it'll try to continue with an empty access token. Not sure if this is the best exception type though.